middle_end/flambda2/tests/algorithms/patricia_tree_tests.ml

```diff
@@ -0,0 +1,915 @@
+    let sort_by_key l = List.sort (fun (k1, _) (k2, _) -> Int.compare k1 k2) l
+
+    let sort_and_group_by_key l =
+      let rec groups l =
+        match l with
+        | [] -> []
+        | (k, v) :: l ->
+          let rec group l =
+            match l with
+            | [] -> [], []
+            | (k', v') :: l when k = k' ->
+              let g, l = group l in
+              v' :: g, l
+            | l -> [], l
+          in
+          let g, l = group l in
+          (k, v :: g) :: groups l
+      in
+      groups (sort_by_key l)
+
+    let same_bindings_up_to_duplicate_keys l1 l2 =
+      let l1 = l1 |> sort_and_group_by_key in
+      let l2 = l2 |> sort_and_group_by_key in
+      let rec check l1 l2 =
+        match l1, l2 with
+        | [], [] -> true
+        | [], _ | _, [] -> false
+        | (k1, vs1) :: l1, (k2, vs2) :: l2 ->
+          Int.equal k1 k2
+          && List.exists (fun v1 -> List.exists (V.equal v1) vs2) vs1
+          && check l1 l2
+      in
+      check l1 l2
+  end
+
+  open Equality_on_bindings
+
+  let of_list_then_bindings l =
+    same_bindings_up_to_duplicate_keys (l |> Map.of_list |> Map.bindings) l
+
+  let bindings_then_of_list m =
+    Map.equal V.equal (m |> Map.bindings |> Map.of_list) m
+
+  (* CR lmaurer: Fix [Tree.Map.disjoint_union] so we can test it *)
+
+  let map_keys_valid f m = Map.valid (Map.map_keys f m)
+
+  let map_keys f m =
+    same_bindings_up_to_duplicate_keys
+      (Map.map_keys f m |> Map.bindings)
+      (List.map (fun (k, v) -> f k, v) (m |> Map.bindings))
+
+  let keys_vs_of_list m =
+    Set.equal (Map.keys m) (m |> Map.bindings |> List.map fst |> Set.of_list)
+
+  let data_vs_bindings m =
+    equal_list_up_to_order V.compare (Map.data m)
+      (Map.bindings m |> List.map snd)
+
+  let of_set_vs_of_list f s =
+    Map.equal V.equal (Map.of_set f s)
+      (Map.of_list (s |> Set.elements |> List.map (fun k -> k, f k)))
+
+  let diff_domains_valid m1 m2 = Map.valid (Map.diff_domains m1 m2)
+
+  let diff_domains m1 m2 =
+    let md = Map.diff_domains m1 m2 in
+    m1
+    |> Map.for_all (fun k v1 ->
+           Map.find_opt k md =? if Map.mem k m2 then None else Some v1)
+
+  let inter_valid f m1 m2 = Map.valid (Map.inter f m1 m2)
+
+  let inter f m1 m2 =
+    let mi = Map.inter f m1 m2 in
+    let in_both k v =
+      match Map.find_opt k m1, Map.find_opt k m2 with
+      | Some v1, Some v2 -> V.equal v (f k v1 v2)
+      | _, _ -> false
+    in
+    Map.for_all in_both mi
+
+  let inter_domain_is_non_empty m1 m2 =
+    Map.inter_domain_is_non_empty m1 m2
+    <=> not (Map.is_empty (Map.inter (fun _k v _ -> v) m1 m2))
+
+  let get_singleton m =
+    match Map.cardinal m with
+    | 1 -> Option.equal equal_bindings (Map.get_singleton m) (Map.choose_opt m)
+    | _ -> Option.is_none (Map.get_singleton m)
+
+  let replace_valid k f m = Map.valid (Map.replace k f m)
+
+  let replace_same k f m =
+    Map.find_opt k (Map.replace k f m) =? Option.map f (Map.find_opt k m)
```

Comment:
How often will `k` be in the map so you can test the `Some` case?  I feel like that'll basically never happen, so you need to constrain the generation of `k` somehow and test both options with separate functions.

Reply:
As above, that should really help.